subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/processing/AnnotationProcessorDetector.java

```diff
@@ -0,0 +1,128 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.tasks.compile.processing;
+
+import com.google.common.base.Charsets;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Lists;
+import com.google.common.io.CharStreams;
+import com.google.common.io.Files;
+import org.apache.tools.zip.ZipEntry;
+import org.apache.tools.zip.ZipFile;
+import org.gradle.api.UncheckedIOException;
+import org.gradle.cache.internal.FileContentCache;
+import org.gradle.cache.internal.FileContentCacheFactory;
+import org.gradle.internal.FileUtils;
+import org.gradle.internal.file.FileType;
+import org.gradle.internal.serialize.ListSerializer;
+import org.gradle.util.DeprecationLogger;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Inspects a classpath to find annotation processors contained in it.
+ */
+public class AnnotationProcessorDetector {
+    private final FileContentCache<List<AnnotationProcessorDeclaration>> cache;
+
+    public AnnotationProcessorDetector(FileContentCacheFactory cacheFactory) {
+        cache = cacheFactory.newCache("annotation-processors", 20000, new ProcessorServiceLocator(), new ListSerializer<AnnotationProcessorDeclaration>(new AnnotationProcessorDeclarationSerializer()));
+    }
+
+    public List<AnnotationProcessorDeclaration> detectProcessors(Iterable<File> processorPath) {
+        List<AnnotationProcessorDeclaration> processors = Lists.newArrayList();
+        for (File jarOrClassesDir : processorPath) {
+            processors.addAll(cache.get(jarOrClassesDir));
+        }
+        return processors;
+    }
+
+    private static class ProcessorServiceLocator implements FileContentCacheFactory.Calculator<List<AnnotationProcessorDeclaration>> {
+        @Override
+        public List<AnnotationProcessorDeclaration> calculate(File file, FileType fileType) {
+            if (fileType == FileType.Directory) {
+                return detectProcessorsInClassesDir(file);
+            }
+            if (fileType == FileType.RegularFile && FileUtils.hasExtensionIgnoresCase(file.getName(), ".jar")) {
+                return detectProcessorsInJar(file);
+            }
+            return Collections.emptyList();
+        }
+
+        private List<AnnotationProcessorDeclaration> detectProcessorsInClassesDir(File classesDir) {
+            File processorDeclaration = new File(classesDir, "META-INF/services/javax.annotation.processing.Processor");
+            if (!processorDeclaration.isFile()) {
+                return Collections.emptyList();
+            }
+            List<String> processorNames = readLines(processorDeclaration);
```

Comment:
:x: this would fail if `javax.annotation.processing.Processor` happens to be a directory. Yeah, I know ;)

Reply:
We check `isFile()` above, which would return false if it's a directory.

Reply:
oops, right